circuit/src/merkle_tree/mod.rs

```diff
@@ -0,0 +1,310 @@
+/// Native version of the chip
+pub mod native;
+
+use std::marker::PhantomData;
+
+use crate::{
+	gadgets::set::{FixedSetChip, FixedSetConfig},
+	params::RoundParams,
+	poseidon::{PoseidonChip, PoseidonConfig},
+};
+use halo2wrong::{
+	curves::FieldExt,
+	halo2::{
+		circuit::{AssignedCell, Layouter, Region},
+		plonk::{Advice, Column, ConstraintSystem, Error},
+	},
+};
+
+const WIDTH: usize = 5;
+
+/// Configuration elements for the circuit are defined here.
+#[derive(Debug, Clone)]
+pub struct MerklePathConfig {
+	/// Configures an advice column for the temp.
+	temp: Column<Advice>,
+	/// Configures FixedSet circuit.
+	set: FixedSetConfig,
+	/// Configures Poseidon circuit.
+	poseidon: PoseidonConfig<WIDTH>,
+}
+
+/// Constructs a chip for the circuit.
+#[derive(Clone)]
+pub struct MerklePathChip<F: FieldExt, const H: usize, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	/// Constructs cell arrays for the nodes.
+	nodes: [AssignedCell<F, F>; H],
+	/// Constructs a phantom data for the parameters.
+	_params: PhantomData<P>,
+}
+
+impl<F: FieldExt, const H: usize, P> MerklePathChip<F, H, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	/// Create a new chip.
+	pub fn new(nodes: [AssignedCell<F, F>; H]) -> Self {
+		MerklePathChip { nodes, _params: PhantomData }
+	}
+
+	/// Make the circuit config.
+	pub fn configure(meta: &mut ConstraintSystem<F>) -> MerklePathConfig {
+		let temp = meta.advice_column();
+		let set = FixedSetChip::<F, 2>::configure(meta);
+		let poseidon = PoseidonChip::<_, WIDTH, P>::configure(meta);
+		meta.enable_equality(temp);
+		MerklePathConfig { temp, set, poseidon }
+	}
+
+	/// Synthesize the circuit.
+	pub fn synthesize(
+		&self, config: MerklePathConfig, path_vec: Vec<F>, zero: AssignedCell<F, F>,
+		mut layouter: impl Layouter<F>,
+	) -> Result<AssignedCell<F, F>, Error> {
+		let mut root: Option<AssignedCell<F, F>> = None;
+		for i in 0..path_vec.len() - 1 {
+			if i % 2 != 0 {
+				continue;
+			}
+			let pos = PoseidonChip::<F, WIDTH, P>::new([
+				self.nodes[i].clone(),
+				self.nodes[i + 1].clone(),
+				zero.clone(),
+				zero.clone(),
+				zero.clone(),
+			]);
+			let hashes =
+				pos.synthesize(&config.poseidon.clone(), layouter.namespace(|| "poseidon"))?;
+			let mut set: Option<FixedSetChip<F, 2>> = None;
+			// When iteration reaches to the root's children this if will trigger and it
+			// will check path_vec has the root inside the set or not. If yes it
+			// will assign the root value and the function will return it.
+			if i == path_vec.len() - 3 {
+				set = Some(FixedSetChip::<F, 2>::new(
+					[path_vec[i + 2], F::zero()],
+					hashes[0].clone(),
+				));
+				root = Some(hashes[0].clone());
+			} else {
+				set = Some(FixedSetChip::<F, 2>::new(
+					[path_vec[i + 2], path_vec[i + 3]],
+					hashes[0].clone(),
+				));
+			}
```

Comment:
I think this path structure is forcing us to write a lot of complicated code. How about changing the struct to accept 2 generic params:
```rust
struct Path<..., const HEIGHT: usize, const ARITY: usize> {
  nodes: [[F; ARITY]; HEIGHT];
}
```

This will force use to make a correct structure before passing it into the chip

Reply:
@lazovicff Is there any significant difference between using an array or a vector in the project? I know that array is more stable and kept at stack of the memory, but is that crucial that much? Because if we use `Vec<Vec<F>>` we won't need `ARITY` and `HEIGHT` constants.

Reply:
We are using arrays because they are more deterministic. We know exactly what to expect and that simplifies the code.
Why do you think its bad to have const generics like ARITY and HEIGHT?

Reply:
If having more const generics are confusing to work with, you can use hardcoded arity of 2

Reply:
No, I am actually comfortable with using const generics, but I thought it will be harder at the readability side. I will focus on using arrays.